### ISSUE TITLE
Telemetry key error

### ISSUE THREAD

**Issue by sleebapaul** [taoxugit/AttnGAN#36]
Hi,
I'm getting the error on running `python main.py --cfg cfg/eval_coco.yml --gpu 1`

```
Traceback (most recent call last):
  File "main.py", line 12, in <module>
    enable(os.environ["TELEMETRY"])
  File "/usr/lib/python2.7/UserDict.py", line 40, in __getitem__
    raise KeyError(key)
KeyError: 'TELEMETRY'
```

**Comment by dc3ea9f** [comment]
The `TELEMETRY` is a environment variable in your system. 

Maybe you can try to run `$ export TELEMETRY=xxx` in bash, then re-run the program.
What's more, the `BLOB_KEY` and `GPU` variable need to be specified too.

**Comment by sleebapaul** [comment]
@MohanZhou Thanks. But where do I get these env variables? The sampling procedure is not mentioning these variables or am I missing something?

**Comment by dc3ea9f** [comment]
As you can see in [`main.py`](https://github.com/taoxugit/AttnGAN/blob/0d000e652b407e976cb88fab299e8566f3de8a37/eval/main.py#L78), and `eval.py` it uses the Azure Storage, you can get key from there. 

But I think it's too silly, in fact, I build and run the docker by disable the Azure service, here is what I got:
```
{
   "bird":{
      "caption":"the bird has a yellow crown and a black eyering that is round",
      "elapsed":10.258235931396484,
      "large":"https://attgan.blob.core.windows.net/images/2018/October/30/11_59_40_514362/attmaps_a1.png",
      "map1":"https://attgan.blob.core.windows.net/images/2018/October/30/11_59_40_514362/bird_g1.png",
      "map2":"https://attgan.blob.core.windows.net/images/2018/October/30/11_59_40_514362/attmaps_a0.png",
      "medium":"https://attgan.blob.core.windows.net/images/2018/October/30/11_59_40_514362/attmaps_a0.png",
      "small":"https://attgan.blob.core.windows.net/images/2018/October/30/11_59_40_514362/bird_g0.png"
   }
}
```
I don't think there exists some useful things for us...

**Comment by sleebapaul** [comment]
@MohanZhou So how we reproduce the repo? Can you guide me to a practical README? You seem to have done a lot of stuff using this repo. Your handle testify.

**Comment by dc3ea9f** [comment]
In this paper, Inception Score and R-precision is used as evaluation metric, the IS is implemented in [this](https://github.com/openai/improved-gan) repository. 

If you want to do R-precision evaluation, you can use the pre-trained DAMSM model to calculate the cosine similarity between global image vectors and the global text vectors. (Refer to the *Section 4.Experiements.Evaluation* of this paper).

**Comment by sleebapaul** [comment]
@MohanZhou I think my intention is much simpler. For now, I want to sample using the pre-trained models of COCO and generate the image for my inputted text. If it works for my application, then I would like to dig deeper. I encountered with env variable error while I tried my cause. 

So can you guide me on this one?

**Comment by dc3ea9f** [comment]
In that case, you should run [`code/main.py`](https://github.com/taoxugit/AttnGAN/blob/master/code/main.py) rather than [`eval/main.py`](https://github.com/taoxugit/AttnGAN/blob/master/eval/main.py).

Refer to the *Sampling* chapter of [`README.md`](https://github.com/taoxugit/AttnGAN/blob/master/README.md).

**Comment by sleebapaul** [comment]
Great! I'll follow that 👍

**Comment by sleebapaul** [comment]
@MohanZhou I ran `python main.py --cfg cfg/eval_coco.yml --data_dir ../data/coco  --gpu 1` from `code/` folder and got the following error. 

```
Load from:  ../data/coco/captions.pickle
Traceback (most recent call last):
  File "main.py", line 130, in <module>
    assert dataset
AssertionError
```

I didn't download the COCO dataset since I don't plan to train. Is that the problem? I've downloaded the pre-trained models for COCO and DAMSSM though.

**Comment by dc3ea9f** [comment]
The problem raised because the `dataset` is `None`. You can just run `main.py` and debug it in the `dataset.py`, find the text data path and fill it.

**Comment by csyan5** [comment]
@sleebapaul Hi Paul. Could I ask are you able to run the program with your customed captions? I couldn't figure out the configuration. Thank you!

**Comment by sleebapaul** [comment]
@ChangshengYan Let me tell you, it's a long road but worth it. I'll explain how I recreated the results in Google Colab. Note that, this is only for testing the repo using pretrained models. I didn't try to train my own. 

#### Install dependencies 
`pip install -q http://download.pytorch.org/whl/{accelerator}/torch-0.4.1-{platform}-linux_x86_64.whl torchvision`
`pip install easydict`
#### Clone the repo
`git clone https://github.com/taoxugit/AttnGAN.git`

#### Move to AttnGAN folder and download some dependencies
```
import os
os.chdir('AttnGAN/data/')
```

#### Trick 1 - Download metadata for COCO
The `coco.zip` link needs to be found by looking closely in README line .
```Download our preprocessed metadata for `birds` `coco` and save them to data/```. 
In this line, birds and coco are different model links to download, but at first look, since the words are close and not seperated by a silly comma, we might get confused a bit. Luckly I hovered the mouse pointer and found that they are seperate links. Once you get the link, download and unzip `coco.zip` and move to `coco` folder.
```
os.chdir("coco")
```

#### Trick 2 - Unzipping val2014-text.zip inside coco folder
This is not explained in README. But you've to unzip this file inside the `coco` folder. Then you need to rename the folder to `text`. The program was continously crashing because it was not able to find the `text` folder. It took me a while to understand this, but renaming was the needed action. 
```
!unzip val2014-text.zip 
!mv val2014/ text
```

#### Download the pretrained models for DAMSM for coco and AttnGAN for coco to respective folders

The links are given in the README. 
```
os.chdir("../../DAMSMencoders/")
```
Download and unzip the `coco.zip`. It'll contain two files `image_encoder100.pth` and `text_encoder100.pth`. Now move to `models` folder.
```
os.chdir("../../models")
```
Download `coco_AttnGAN2.pth` to this folder and move back to `code` folder.
```
os.chdir("../code")
```

#### Trick 3 - Sampling
Though the authors have mentioned that sampling can be done using `main.py` command, but not which `main.py`. Actually it is the `main.py` in the `code` folder. Thanks to @MohanZhou for helping me finding it. 

### Trick 4 - Font issue of PIL
This might not be an issue for you if you're not using Colab, but I had real bad time fixing this issue. PIL font path mentioned in the `miscc/utils.py` was not working. It took me 1-2 hours to find the issue and I re-wrote a line in `miscc/utils.py`. If you run into the same issue, do the following. 

1. Comment the following line in `miscc/utils.py`
```
fnt = ImageFont.truetype('Pillow/Tests/fonts/FreeMono.ttf', 50)
```
2. Replace it with 
```
fnt = ImageFont.load_default()
```

#### Now you're good to go

Add your custom text in `data/coco/example_captions.txt` and move to `code` folder. 
```
os.chdir("../../code")
```
Run the command. If you've a single GPU, then GPU id will be `0`. 
`!python main.py --cfg cfg/eval_coco.yml --gpu 0`

#### Trick 5 - Finding the generated images 
I've had a bad time finding the generated samples as well. Finally I found out the output path as,
`../models/coco_AttnGAN2/example_captions/`

So in a nutshell, the repo is great and served my purpose well. But I wish the authors had put their best efforts while writing the README as well.

**Comment by csyan5** [comment]
@sleebapaul Thanks, Paul! I tried to run on CPU and it now can generate images from text from customed inputs. Your instructions are really helpful. Thank you!

**Comment by ghost** [comment]
@sleebapaul do you have the colab notebook, I would like to try it out thanks
